### ISSUE TITLE
Regression in destructuring models from effects dispatch param

### ISSUE THREAD

**Issue by semoal** [rematch/rematch#811]
[<img alt="Issuehunt badges" src="https://img.shields.io/badge/IssueHunt-%2425.8%20Funded-%2300A156.svg" />](https://issuehunt.io/r/rematch/rematch/issues/811)

**Describe the bug**

Regression about destructuring dispatch models from function param.

**To Reproduce**
```js
export const count = {
	state: 0,
	reducers: {
		increment: s => s + 1,
	},
	effects: ({ count }) => ({
		async asyncIncrement() {
			await new Promise(resolve => {
				setTimeout(resolve, 1000)
			})

			count.increment()
		},
	}),
}
```

**Expected behavior**
Should work on the same way than not destructured like v1.

**Screenshots**
<img width="933" alt="Screenshot 2020-08-27 at 22 34 53" src="https://user-images.githubusercontent.com/22656541/91492382-d8aa8580-e8b5-11ea-87a3-a759c32b93a6.png">

**Additional context**
There's an issue related to circular reference that @tianzhich is looking for a solution, could be related.
Destructuring the models inside the effect function works good.
```js
export const count = {
	state: 0,
	reducers: {
		increment: s => s + 1,
	},
	effects: dispatch => ({
		async asyncIncrement() {
			const { count } = dispatch
			count.increment()
		},
	}),
}
```

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Backers (Total: $25.80)

- [<img src='https://avatars.githubusercontent.com/u/22656541?v=4' alt='semoal' width=24 height=24> semoal](https://issuehunt.io/u/semoal) ($25.80)

### Submitted pull Requests
- [#947 fix: circular reference destruction works with models](https://issuehunt.io/r/rematch/rematch/pull/947)
---

### Tips

- Checkout the [Issuehunt explorer](https://issuehunt.io/r/rematch/rematch/) to discover more funded issues.
- Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on IssueHunt to raise funds.
</details>

**Comment by tianzhich** [comment]
related https://github.com/microsoft/TypeScript/issues/40279

**Comment by tianzhich** [comment]
Hi, @pjamrozowicz I have found 2 related problems with this issue, it's not only about typescript but also javascript.

We are waiting for TypeScript to fix the ts problem. But about js, Look at the snippet below:

We `createDispatcher` like this, 

![image](https://user-images.githubusercontent.com/33218310/92191339-1d29b880-ee96-11ea-9065-f5183af5678d.png)

And inside it, pass the `rematch.dispatch` to `modelEffectsCreator`, 

![image](https://user-images.githubusercontent.com/33218310/92191396-43e7ef00-ee96-11ea-8a14-0cb4d97441ac.png)

This time we cant access `rematch.dispatch.currentModel` cause the dispatcher for this model wasn't created till now.

So we got `undefined` https://github.com/rematch/rematch/issues/811#issue-687525587 while effects running.

There was a closure around the returned effects object. If it referenced to dispatch directly,  all would be ok. But if it referenced to the destructed `{ count } = dispatch`, it will reference to undefined.

**Comment by tianzhich** [comment]
I seem to figure it out a little bit. 

In the v1 version, Shawn split the reducers creator and effects creator into 2 separate plugins and creating effects after reducers created. So the effects creator can access the model cause it was created in reducer creation. 

But we combine them into createDispatcher in v2, and not inject created modelDispacther to `rematch.dispatch` before it was passed to effectsCreator.

I modify the code like following and resolve this:

![image](https://user-images.githubusercontent.com/33218310/92198529-3424d680-eea7-11ea-9317-8cc2ecea71ae.png)

**Comment by tianzhich** [comment]
@semoal We still can't destruct dispatch from param of effectsCreator. I think it was a pending issue that TypeScript should fix for us. There is no difference between destruction from param or inside the function. 

But now it functions normally and is consistent with v1.☺️

**Comment by Hainesy** [comment]
I think this is the same issue I raised here: https://github.com/rematch/rematch/issues/760

I have experienced it when there is a circular dependency between two models, i.e. effects from each model are dependent on the other.

**Comment by semoal** [comment]
> I think this is the same issue I raised here: #760
> 
> I have experienced it when there is a circular dependency between two models, i.e. effects from each model are dependent on the other.

we released 2.0.0-next.5, could you try it out? in some internal codebases this issue is fixed. (we still have the typescript issue)

**Comment by semoal** [comment]
I close this issue cuz it's fixed the circular referencing issue on javascript codebases.
I'll open a new one to track typescript issue
